### PR TITLE
GC Tombstone: Separate flags for Load v. Usage (required for recovery)

### PR DESCRIPTION
**Pull request by markfields**
## Description

We are working on a mechanism to allow consumers to retrieve and revive a Tombstoned object (see #13606). In these cases, the object will still appear tombstoned so we need to allow tombstoned objects to be used (ops sent/received).

This PR separates two flags where there was previously one.  Now it's `Fluid.GarbageCollection.ThrowOnTombstoneUsage` and `Fluid.GarbageCollection.ThrowOnTombstoneLoad`

## Reviewer Guidance

I also made some changes to `sendGCTombstoneEvent` in here.

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 1]
```diff
@@ -186,6 +188,7 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
             assert(response2.value.startsWith("Blob removed by gc:"), `Unexpected response value for blob handle 2`);
         });
 
+        //* Update this
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:191]:
```suggestion
```

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 2]
```diff
@@ -548,6 +549,7 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
             assert(response.value === `Datastore removed by gc: ${unreferencedId}`);
         });
 
+        //* Update this
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:552]:
```suggestion
```

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 3]
```diff
@@ -444,19 +444,19 @@ export class DataStores implements IDisposable {
             // Note: if a user writes a request to look like it's viaHandle, we will also send this telemetry event
             const event = {
                 eventName: "GC_Tombstone_DataStore_Requested",
-                url: request.url,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStores.ts:447]: The URL is already included in the error message (in a sanitized form)

File: packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts [hunk 4]
```diff
@@ -6,26 +6,24 @@
...
- * Decides whether or not to send an error event or a generic event for gc tombstone scenarios
- *
- * Adds isSummarizerClient, packagePath, and error to telemetry properties.
+ * Consolidates info / logic for logging when we encounter a Tombstone
  */
 export function sendGCTombstoneEvent(
     mc: MonitoringContext,
-    event: ITelemetryGenericEvent,
-    isSummarizerClient: boolean,
+    event: ITelemetryGenericEvent & { isSummarizerClient: boolean },
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts:16]: Why remove this as a separate param and add to the event?

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 5]
```diff
@@ -58,8 +58,9 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
         if (provider.driver.type !== "local") {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:58]: There should be tests that validate usage of this. I believe your [other PR](https://github.com/microsoft/FluidFramework/pull/13606) will have it but mentioning here.

**Review by agarwal-navin — APPROVED** [review]

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 5 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:58]: Yeah let me add tests here for this change itself

File: packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts (hunk 4 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts:16]: I was thinking about your feedback in the other PR (similar change), I think you were suggesting to take in all the props directly rather than taking in an event. But then we'd need a generic property bag, and the result will be a longer and longer function signature which is a smell to me.  So I tried to consolidate as much data as possible in the event itself (which also supports arbitrary other props), and only have "top-level" function params for data that needs to be interpreted/transformed.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 5 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:58]: @agarwal-navin done

File: packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts [hunk 6]
```diff
@@ -6,26 +6,23 @@
...
-    event: ITelemetryGenericEvent,
-    isSummarizerClient: boolean,
+    event: ITelemetryGenericEvent & { category: "error" | "generic", isSummarizerClient: boolean },
     packagePath: readonly string[] | undefined,
-    error?: any,
+    error?: unknown,
 ) {
-    const throwOnTombstoneUsage = event.throwOnTombstoneUsage = mc.config.getBoolean(throwOnTombstoneUsageKey) ?? false;
     event.pkg = packagePathToTelemetryProperty(packagePath);
-    event.isSummarizerClient = isSummarizerClient;
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts:25]: Why was this removed?
